tests/Unit/Product/ProductRepositoryTest.php

```diff
@@ -0,0 +1,272 @@
+					'compare' => '=',
+					'value'   => '',
+				],
+			],
+		];
+		$this->assertEquals( [ $product_1->get_id() ], $this->product_repository->find_ids( $query_args ) );
+		$this->assertEquals( [ wc_get_product( $product_1 ) ], $this->product_repository->find( $query_args ) );
+	}
+
+	public function test_find_by_ids() {
+		$product_1 = WC_Helper_Product::create_simple_product();
+		$product_2 = WC_Helper_Product::create_simple_product();
+
+		$ids = [
+			$product_1->get_id(),
+			$product_2->get_id(),
+		];
+
+		$this->assertEqualSets(
+			array_map( 'wc_get_product', $ids ),
+			$this->product_repository->find_by_ids( $ids )
+		);
+	}
+
+	public function test_find_synced_products() {
+		$product_1 = WC_Helper_Product::create_simple_product();
+		$this->product_helper->mark_as_synced( $product_1, $this->generate_google_product_mock() );
+
+		WC_Helper_Product::create_simple_product();
+
+		$this->assertEquals(
+			[ wc_get_product( $product_1->get_id() ) ],
+			$this->product_repository->find_synced_products()
+		);
+
+		$this->assertEquals(
+			[ $product_1->get_id() ],
+			$this->product_repository->find_synced_product_ids()
+		);
+	}
+
+	public function test_find_sync_ready_products() {
+		// create some products that are not sync ready
+
+		// manually set to dont-sync-and-show
+		$product_1 = WC_Helper_Product::create_simple_product();
+		$this->product_meta->update_visibility( $product_1, ChannelVisibility::DONT_SYNC_AND_SHOW );
+
+		// product status is set to private and therefore not supported
+		$product_2 = WC_Helper_Product::create_simple_product();
+		$this->product_meta->update_visibility( $product_2, ChannelVisibility::SYNC_AND_SHOW );
+		$product_2->set_status( 'private' );
+		$product_2->save();
+
+		// product type is not supported
+		$product_3 = WC_Helper_Product::create_external_product();
+		$this->product_meta->update_visibility( $product_3, ChannelVisibility::SYNC_AND_SHOW );
+
+		// syncing this product has failed several times recently
+		$product_4 = WC_Helper_Product::create_simple_product();
+		$this->product_meta->update_visibility( $product_4, ChannelVisibility::SYNC_AND_SHOW );
+		$this->product_meta->update_failed_sync_attempts( $product_4, ProductSyncer::FAILURE_THRESHOLD + 1 );
+		$this->product_meta->update_sync_failed_at( $product_4, strtotime( '+1 year' ) );
+
+		// a simple product that is sync ready
+		$simple_product = WC_Helper_Product::create_simple_product();
+		$this->product_meta->update_visibility( $simple_product, ChannelVisibility::SYNC_AND_SHOW );
+
+		// a variable product that is sync ready along with all its variations
+		$variable_product = WC_Helper_Product::create_variation_product();
+		$this->product_meta->update_visibility( wc_get_product( $variable_product ), ChannelVisibility::SYNC_AND_SHOW );
+		foreach ( $variable_product->get_children() as $variation_id ) {
+			$this->product_meta->update_visibility( wc_get_product( $variation_id ), ChannelVisibility::SYNC_AND_SHOW );
+		}
```

Comment:
As far as I know we don't set this for each variation when updating a product through the ChannelVisibilityMetaBox. Although we might want to consider doing so, as it can lead to creating incomplete batches when syncing products. Was still planning to log an issue for this.

Reply:
Here is the issue: https://github.com/woocommerce/google-listings-and-ads/issues/861